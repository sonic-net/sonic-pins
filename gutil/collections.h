// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef GUTIL_COLLECTIONS_H
#define GUTIL_COLLECTIONS_H

#include <string>
#include <type_traits>
#include <vector>

#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "glog/logging.h"
#include "google/protobuf/map.h"
#include "gutil/status.h"

namespace gutil {

// Returns a vector of the keys in a map collection.
template <typename M>
std::vector<typename M::key_type> Keys(const M &m) {
  std::vector<typename M::key_type> keys;
  for (const auto &[key, val] : m) keys.push_back(key);
  return keys;
}

// Returns a const copy of the value associated with a given key if it exists,
// or a status failure if it does not.
//
// WARNING: prefer FindOrNull if the value can be large to avoid the copy.
template <typename M, typename KeyType = typename M::key_type>
absl::StatusOr<const typename M::mapped_type> FindOrStatus(const M &m,
                                                           const KeyType &k) {
  auto it = m.find(k);
  if (it != m.end()) return it->second;
  if constexpr (std::is_same_v<KeyType, std::string>) {
    return absl::NotFoundError(absl::StrCat("Key not found: '", k, "'"));
  } else {
    return absl::NotFoundError("Key not found");
  }
}

// Returns a non-null pointer of the value associated with a given key
// if it exists, or a status failure if it does not.
template <typename M, typename KeyType = typename M::key_type>
absl::StatusOr<const typename M::mapped_type *> FindPtrOrStatus(
    M &m, const KeyType &k) {
  auto it = m.find(k);
  if (it != m.end()) return &it->second;
  if constexpr (std::is_same_v<typename M::key_type, std::string>) {
    return absl::NotFoundError(absl::StrCat("Key not found: '", k, "'"));
  } else {
    return absl::NotFoundError("Key not found");
  }
}

// Returns a const pointer of the value associated with a given key if it
// exists, or a nullptr if it does not.
template <typename M, typename KeyType = typename M::key_type>
const typename M::mapped_type *FindOrNull(const M &m, const KeyType &k) {
  const auto it = m.find(k);
  if (it != m.end()) return &(it->second);
  return nullptr;
}

// Returns a non-const pointer of the value associated with a given key if it
// exists, or a nullptr if it does not.
template <typename M, typename KeyType = typename M::key_type>
typename M::mapped_type *FindOrNull(M &m, const KeyType &k) {
  auto it = m.find(k);
  if (it != m.end()) return &(it->second);
  return nullptr;
}

// Returns a reference of the value associated with a given key if it exists,
// crashes if it does not.
template <typename M, typename KeyType = typename M::key_type>
typename M::mapped_type &FindOrDie(M &map, const KeyType &key) {
  auto iter = map.find(key);
  CHECK(iter != map.end()) << "Could not find key";  // Crash OK
  return iter->second;
}

// Returns a const reference of the value associated with a given key if it
// exists, crashes if it does not.
template <typename M, typename KeyType = typename M::key_type>
const typename M::mapped_type &FindOrDie(const M &map, const KeyType &key) {
  auto iter = map.find(key);
  CHECK(iter != map.end()) << "Could not find key";  // Crash OK

  return iter->second;
}

// Returns a copy of the value associated with the given key if it exists, or
// returns the given default value otherwise.
// NOTE: May be inefficient for large datatypes that are expensive to copy.
template <typename M, typename KeyType = typename M::key_type>
typename M::mapped_type FindOrDefault(const M &map, const KeyType &key,
                                      typename M::mapped_type default_value) {
  auto it = map.find(key);
  return (it != map.end()) ? it->second : default_value;
}

// Checks if the id is unique in set.
template <typename M>
absl::Status InsertIfUnique(absl::flat_hash_set<M> &set, const M &id,
                            const std::string &error_message) {
  const auto it = set.insert(id);
  if (!it.second) {
    return absl::Status(absl::StatusCode::kInvalidArgument, error_message);
  }

  return absl::OkStatus();
}

template <typename K, typename V>
absl::Status InsertIfUnique(absl::flat_hash_map<K, V> &map, K key, const V &val,
                            const std::string &error_message) {
  auto it = map.insert({key, val});
  if (!it.second) {
    return absl::Status(absl::StatusCode::kInvalidArgument, error_message);
  }

  return absl::OkStatus();
}

template <typename K, typename V>
absl::Status InsertIfUnique(google::protobuf::Map<K, V> *map, K key,
                            const V &val, const std::string &error_message) {
  auto it = map->insert({key, val});
  if (!it.second) {
    return absl::Status(absl::StatusCode::kInvalidArgument, error_message);
  }

  return absl::OkStatus();
}

}  // namespace gutil

#endif  // GUTIL_COLLECTIONS_H
