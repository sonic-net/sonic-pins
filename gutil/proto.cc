// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "gutil/proto.h"

#include <fcntl.h>

#include <string>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "google/protobuf/descriptor.h"
#include "google/protobuf/io/zero_copy_stream_impl.h"
#include "google/protobuf/message.h"
#include "google/protobuf/text_format.h"
#include "gutil/status.h"

namespace gutil {

bool IsEmptyProto(const google::protobuf::Message &message) {
  return message.ByteSizeLong() == 0;
}

absl::Status ReadProtoFromFile(absl::string_view filename,
                               google::protobuf::Message *message) {
  // Verifies that the version of the library that we linked against is
  // compatible with the version of the headers we compiled against.
  GOOGLE_PROTOBUF_VERIFY_VERSION;

  int fd = open(std::string(filename).c_str(), O_RDONLY);
  if (fd < 0) {
    return InvalidArgumentErrorBuilder()
           << "Error opening the file " << filename;
  }

  google::protobuf::io::FileInputStream file_stream(fd);
  file_stream.SetCloseOnDelete(true);

  if (!google::protobuf::TextFormat::Parse(&file_stream, message)) {
    return InvalidArgumentErrorBuilder() << "Failed to parse file " << filename;
  }

  return absl::OkStatus();
}

absl::Status ReadProtoFromString(absl::string_view proto_string,
                                 google::protobuf::Message *message) {
  // Verifies that the version of the library that we linked against is
  // compatible with the version of the headers we compiled against.
  GOOGLE_PROTOBUF_VERIFY_VERSION;

  if (!google::protobuf::TextFormat::ParseFromString(std::string(proto_string),
                                                     message)) {
    return InvalidArgumentErrorBuilder()
           << "Failed to parse string " << proto_string;
  }

  return absl::OkStatus();
}

absl::Status SaveProtoToFile(absl::string_view filename,
                             const google::protobuf::Message &message) {
  // Verifies that the version of the library that we linked against is
  // compatible with the version of the headers we compiled against.
  GOOGLE_PROTOBUF_VERIFY_VERSION;
  int fd = open(std::string(filename).c_str(), O_WRONLY | O_CREAT, 0644);
  if (fd < 0) {
    return InvalidArgumentErrorBuilder()
           << "Error opening the file " << filename;
  }

  google::protobuf::io::FileOutputStream file_stream(fd);
  file_stream.SetCloseOnDelete(true);

  if (!google::protobuf::TextFormat::Print(message, &file_stream)) {
    return InvalidArgumentErrorBuilder()
           << "Failed to print proto to file " << filename;
  }

  file_stream.Flush();
  return absl::OkStatus();
}

absl::StatusOr<std::string> ProtoDiff(
    const google::protobuf::Message &message1,
    const google::protobuf::Message &message2,
    google::protobuf::util::MessageDifferencer differ) {
  if (message1.GetDescriptor() != message2.GetDescriptor()) {
    return gutil::InvalidArgumentErrorBuilder()
           << "cannot compute diff for messages of incompatible descriptors `"
           << message1.GetDescriptor()->full_name() << "' vs '"
           << message2.GetDescriptor()->full_name() << "'";
  }

  std::string diff;
  differ.ReportDifferencesToString(&diff);
  differ.Compare(message1, message2);
  return diff;
}

absl::StatusOr<std::string> GetOneOfFieldName(
    const google::protobuf::Message &message, const std::string &oneof_name) {
  const auto *oneof_descriptor =
      message.GetDescriptor()->FindOneofByName(oneof_name);
  const auto *field = message.GetReflection()->GetOneofFieldDescriptor(
      message, oneof_descriptor);
  if (!field) {
    return gutil::NotFoundErrorBuilder()
           << "Oneof field \"" << oneof_name << "\" is not set";
  }
  return field->name();
}

std::string PrintTextProto(const google::protobuf::Message &message) {
  std::string message_text;

  google::protobuf::TextFormat::Printer printer;
  printer.SetExpandAny(true);

  printer.PrintToString(message, &message_text);

  return message_text;
}

// Print proto in TextFormat with single line mode enabled.
std::string PrintShortTextProto(const google::protobuf::Message &message) {
  std::string message_short_text;

  google::protobuf::TextFormat::Printer printer;
  printer.SetSingleLineMode(true);
  printer.SetExpandAny(true);

  printer.PrintToString(message, &message_short_text);
  // Single line mode currently might have an extra space at the end.
  if (!message_short_text.empty() && message_short_text.back() == ' ') {
    message_short_text.pop_back();
  }

  return message_short_text;
}

absl::StatusOr<std::string> SerializeProtoAsJson(
    const google::protobuf::Message &proto) {
  std::string json;
  RETURN_IF_ERROR(gutil::ToAbslStatus(
      google::protobuf::util::MessageToJsonString(proto, &json)));
  return json;
}

}  // namespace gutil
