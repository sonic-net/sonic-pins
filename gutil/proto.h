// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef GUTIL_PROTO_H
#define GUTIL_PROTO_H

#include <string>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "google/protobuf/message.h"
#include "google/protobuf/util/json_util.h"
#include "google/protobuf/util/message_differencer.h"
#include "gutil/status.h"

namespace gutil {

// Returns `true` if the given `message` has no fields set, `false` otherwise.
bool IsEmptyProto(const google::protobuf::Message &message);

// Read the contents of the file into a protobuf.
absl::Status ReadProtoFromFile(absl::string_view filename,
                               google::protobuf::Message *message);

// Read the contents of the string into a protobuf.
absl::Status ReadProtoFromString(absl::string_view proto_string,
                                 google::protobuf::Message *message);

// Saves the content of a protobuf into a file.
absl::Status SaveProtoToFile(absl::string_view filename,
                             const google::protobuf::Message &message);

// Read the contents of the given string into a protobuf and returns it.
template <class T>
absl::StatusOr<T> ParseTextProto(absl::string_view proto_string) {
  T message;
  RETURN_IF_ERROR(ReadProtoFromString(proto_string, &message));
  return message;
}

// Returns diff of the given protobuf messages, provided they have the same
// `Descriptor` (message1.GetDescriptor() == message2.GetDescriptor()), or
// returns `InvalidArgumentError` otherwise.
// Optionally, a `differ` can be provided for fine-grained control over how to
// compute the diff.
absl::StatusOr<std::string> ProtoDiff(
    const google::protobuf::Message &message1,
    const google::protobuf::Message &message2,
    google::protobuf::util::MessageDifferencer differ =
        google::protobuf::util::MessageDifferencer());

// Get the name of the oneof enum that is set.
// Eg:
// message IrValue {
//   oneof format {
//     string hex_str = 1;
//     string ipv4 = 2;
//     string ipv6 = 3;
//     string mac = 4;
//     string str = 5;
//   }
// }
// IrValue value;
// value.set_hex_str("0xf00d");
// std::string name = GetOneOfFieldName(value, std::string("format"));
// EXPECT_EQ(name, "hex_str");
absl::StatusOr<std::string> GetOneOfFieldName(
    const google::protobuf::Message &message, const std::string &oneof_name);

// Print proto in TextFormat.
std::string PrintTextProto(const google::protobuf::Message &message);

// Print proto in TextFormat in a single line.
std::string PrintShortTextProto(const google::protobuf::Message &message);

// Parses the given JSON string into a proto of type `T`.
template <class T>
absl::StatusOr<T> ParseJsonAsProto(absl::string_view raw_json_string,
                                   bool ignore_unknown_fields = false);

// Serializes the given proto message as a JSON string.
absl::StatusOr<std::string> SerializeProtoAsJson(
    const google::protobuf::Message &proto);

// -- END OF PUBLIC INTERFACE - Implementation details follow ------------------

template <class T>
absl::StatusOr<T> ParseJsonAsProto(absl::string_view raw_json_string,
                                   bool ignore_unknown_fields) {
  google::protobuf::util::JsonParseOptions options;
  options.ignore_unknown_fields = ignore_unknown_fields;
  T proto;
  // OS protobuf uses its own `Status`-like and `string_view`-like classes, so
  // some gymnastics are required here:
  // - ToAbslStatus converts any `Status`-like type to an absl::Status.
  // - We pass in `{raw_json_string.data(), raw_json_string.size()}` instead of
  // `raw_json_string`, constructing a new object of the appropriate
  // `string_view`-like type implicitly.
  RETURN_IF_ERROR(
      gutil::ToAbslStatus(google::protobuf::util::JsonStringToMessage(
          {raw_json_string.data(), raw_json_string.size()}, &proto, options)));
  return proto;
}

}  // namespace gutil

#endif  // GUTIL_PROTO_H
