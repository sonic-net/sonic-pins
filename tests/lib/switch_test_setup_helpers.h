#ifndef PINS_TESTS_LIB_SWITCH_TEST_SETUP_HELPERS_H_
#define PINS_TESTS_LIB_SWITCH_TEST_SETUP_HELPERS_H_

#include <memory>
#include <optional>
#include <string>
#include <utility>
#include <vector>

#include "absl/container/btree_set.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/time/time.h"
#include "absl/types/span.h"
#include "lib/gnmi/openconfig.pb.h"
#include "p4/config/v1/p4info.pb.h"
#include "p4_pdpi/ir.pb.h"
#include "p4_pdpi/p4_runtime_session.h"
#include "thinkit/mirror_testbed.h"
#include "thinkit/switch.h"

namespace pins_test {

// Configures the switch and sets up a P4 Runtime Session. If you don't have
// particular requirements, this is likely the function you want to use.
// Specifically:
// * Creates a session.
// * Clears all tables.
// * Pushes the given `gnmi_config`, if any, and waits for the switch to
//   converge.
// * Pushes the given P4Info, if any, via RECONCILE_AND_COMMIT.
absl::StatusOr<std::unique_ptr<pdpi::P4RuntimeSession>>
ConfigureSwitchAndReturnP4RuntimeSession(
    thinkit::Switch &thinkit_switch, std::optional<std::string> gnmi_config,
    std::optional<p4::config::v1::P4Info> p4info,
    const pdpi::P4RuntimeSessionOptionalArgs &metadata = {});

// Configures a pair of switches and sets up P4 Runtime Sessions. If you are
// setting up a pair of switches (e.g. in a mirror testbed) with the same gNMI
// config and P4Info and want to wait for the gNMI configs to converge in
// parallel, this is likely the function that you should use. Specifically, it:
// * Creates two sessions.
// * Clears all tables.
// * Pushes the given `gnmi_config`, if any, and waits for the switches to
//   converge.
// * Pushes the given `p4info`, if any,  via RECONCILE_AND_COMMIT.
absl::StatusOr<std::pair<std::unique_ptr<pdpi::P4RuntimeSession>,
                         std::unique_ptr<pdpi::P4RuntimeSession>>>
ConfigureSwitchPairAndReturnP4RuntimeSessionPair(
    thinkit::Switch &thinkit_switch1, thinkit::Switch &thinkit_switch2,
    std::optional<std::string> gnmi_config,
    std::optional<p4::config::v1::P4Info> p4info,
    const pdpi::P4RuntimeSessionOptionalArgs &metadata = {});

// Mirrors the P4RT port IDs of the SUT onto the control switch and ensures that
// the state of the switch is properly updated (i.e. that the IDs are reflected
// in the gNMI state path).
absl::Status MirrorSutP4rtPortIdConfigToControlSwitch(
    thinkit::MirrorTestbed &testbed,
    absl::Duration config_convergence_timeout_per_switch = absl::Minutes(5));

// Reads the enabled interfaces from the switch and waits up to `timeout` until
// they are all up. Calls `on_failure` prior to returning status if it is not
// OK.
absl::Status WaitForEnabledInterfacesToBeUp(
    thinkit::Switch &thinkit_switch, absl::Duration timeout = absl::Minutes(5),
    std::optional<
        std::function<void(const openconfig::Interfaces &enabled_interfaces)>>
        on_failure = std::nullopt);

// Gets a count of which P4runtime ports are used in `entries` and how
// frequently.
absl::StatusOr<absl::btree_set<std::string>>
GetPortsUsed(const pdpi::IrP4Info &info,
             std::vector<pdpi::IrTableEntry> entries);

// Rewrites the ports of the given table `entries` to only use the given
// non-empty `ports`. Uses `info` to determine which values refer to ports.
// Specifically, for each port `x` in the set of table entries, this
// function remaps it to f(x) such that:
// - if `x \in ports` then `f(x) = x`
// - `x` is remapped to an `f(x) \in ports` (chosen deterministically)
//   such that `\forall. p1,p2 \in ports. |{f(x) = p1 | x \in
//   all_ports(entries)}| <= |{f(x) = p2 | x \in all_ports(entries)}| + 1`.
//
// This function makes it possible to use a list of pre-existing table entries
// with any set of ports desired (or configured on the switch).
absl::Status
RewritePortsInTableEntries(const pdpi::IrP4Info &info,
                           absl::Span<const std::string> new_ports,
                           std::vector<pdpi::IrTableEntry> &entries);

// Extracts the available ports from the given `gnmi_config` and rewrites
// entries as per above.
absl::Status
RewritePortsInTableEntries(const pdpi::IrP4Info &info,
                           absl::string_view gnmi_config,
                           std::vector<pdpi::IrTableEntry> &entries);

} // namespace pins_test

#endif // PINS_TESTS_LIB_SWITCH_TEST_SETUP_HELPERS_H_
