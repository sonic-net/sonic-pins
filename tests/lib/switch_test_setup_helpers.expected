================================================================================
1,2,some_port -> a_port
Available ports: a_port
================================================================================
-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "router_interface_table"
matches {
  name: "router_interface_id"
  exact {
    str: "router-interface-1"
  }
}
action {
  name: "set_port_and_src_mac"
  params {
    name: "port"
    value {
      str: "1"
    }
  }
  params {
    name: "src_mac"
    value {
      mac: "00:02:03:04:05:06"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action.params[0].value.str: "1" -> "a_port"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "wcmp_group_table"
matches {
  name: "wcmp_group_id"
  exact {
    str: "group-4294934545"
  }
}
action_set {
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-1"
        }
      }
    }
    weight: 1
    watch_port: "2"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "some_port"
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action_set.actions[0].watch_port: "2" -> "a_port"
modified: action_set.actions[1].watch_port: "some_port" -> "a_port"

================================================================================
1 -> Exactly one of {2,3}
Available ports: 2,3
================================================================================
-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "router_interface_table"
matches {
  name: "router_interface_id"
  exact {
    str: "router-interface-1"
  }
}
action {
  name: "set_port_and_src_mac"
  params {
    name: "port"
    value {
      str: "1"
    }
  }
  params {
    name: "src_mac"
    value {
      mac: "00:02:03:04:05:06"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action.params[0].value.str: "1" -> "3"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "wcmp_group_table"
matches {
  name: "wcmp_group_id"
  exact {
    str: "group-4294934545"
  }
}
action_set {
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-1"
        }
      }
    }
    weight: 1
    watch_port: "1"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "1"
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action_set.actions[0].watch_port: "1" -> "3"
modified: action_set.actions[1].watch_port: "1" -> "3"

================================================================================
Unchanged due to only existing ports
Available ports: 2,3
================================================================================
-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "router_interface_table"
matches {
  name: "router_interface_id"
  exact {
    str: "router-interface-1"
  }
}
action {
  name: "set_port_and_src_mac"
  params {
    name: "port"
    value {
      str: "2"
    }
  }
  params {
    name: "src_mac"
    value {
      mac: "00:02:03:04:05:06"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "wcmp_group_table"
matches {
  name: "wcmp_group_id"
  exact {
    str: "group-4294934545"
  }
}
action_set {
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-1"
        }
      }
    }
    weight: 1
    watch_port: "2"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "3"
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

================================================================================
All unchanged due to no ports
Available ports: 1
================================================================================
-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "neighbor_table"
matches {
  name: "router_interface_id"
  exact {
    str: "router-interface-1"
  }
}
matches {
  name: "neighbor_id"
  exact {
    str: "fe80::3e28:6dff:fe34:c002"
  }
}
action {
  name: "set_dst_mac"
  params {
    name: "dst_mac"
    value {
      mac: "3c:28:6d:34:c0:02"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "nexthop_table"
matches {
  name: "nexthop_id"
  exact {
    str: "nexthop-1"
  }
}
action {
  name: "set_ip_nexthop"
  params {
    name: "router_interface_id"
    value {
      str: "router-interface-1"
    }
  }
  params {
    name: "neighbor_id"
    value {
      str: "fe80::3e28:6dff:fe34:c002"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "ipv4_table"
matches {
  name: "vrf_id"
  exact {
    str: "vrf-80"
  }
}
action {
  name: "set_wcmp_group_id"
  params {
    name: "wcmp_group_id"
    value {
      str: "group-4294934531"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "ipv6_table"
matches {
  name: "vrf_id"
  exact {
    str: "vrf-80"
  }
}
action {
  name: "set_wcmp_group_id"
  params {
    name: "wcmp_group_id"
    value {
      str: "group-4294934531"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "vrf_table"
matches {
  name: "vrf_id"
  exact {
    str: "vrf-80"
  }
}
action {
  name: "no_action"
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

================================================================================
Roughly even number of 1, 2, and 'a_port', with existing 'a_port' and port-less entries unchanged
Available ports: 1,2,a_port
================================================================================
-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "neighbor_table"
matches {
  name: "router_interface_id"
  exact {
    str: "router-interface-1"
  }
}
matches {
  name: "neighbor_id"
  exact {
    str: "fe80::3e28:6dff:fe34:c002"
  }
}
action {
  name: "set_dst_mac"
  params {
    name: "dst_mac"
    value {
      mac: "3c:28:6d:34:c0:02"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "router_interface_table"
matches {
  name: "router_interface_id"
  exact {
    str: "router-interface-1"
  }
}
action {
  name: "set_port_and_src_mac"
  params {
    name: "port"
    value {
      str: "a_port"
    }
  }
  params {
    name: "src_mac"
    value {
      mac: "00:02:03:04:05:06"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
NONE

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "wcmp_group_table"
matches {
  name: "wcmp_group_id"
  exact {
    str: "group-4294934545"
  }
}
action_set {
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-1"
        }
      }
    }
    weight: 1
    watch_port: "4"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "5"
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action_set.actions[0].watch_port: "4" -> "a_port"
modified: action_set.actions[1].watch_port: "5" -> "2"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "mirror_session_table"
matches {
  name: "mirror_session_id"
  exact {
    str: "mirror-session-201326593"
  }
}
action {
  name: "mirror_as_ipv4_erspan"
  params {
    name: "port"
    value {
      str: "6"
    }
  }
  params {
    name: "src_ip"
    value {
      ipv4: "10.206.196.0"
    }
  }
  params {
    name: "dst_ip"
    value {
      ipv4: "172.20.0.202"
    }
  }
  params {
    name: "src_mac"
    value {
      mac: "00:02:03:04:05:06"
    }
  }
  params {
    name: "dst_mac"
    value {
      mac: "00:1a:11:17:5f:80"
    }
  }
  params {
    name: "ttl"
    value {
      hex_str: "0x40"
    }
  }
  params {
    name: "tos"
    value {
      hex_str: "0x00"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action.params[0].value.str: "6" -> "1"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "l3_admit_table"
matches {
  name: "dst_mac"
  ternary {
    value {
      mac: "02:32:0a:ce:c4:04"
    }
    mask {
      mac: "ff:ff:ff:ff:ff:ff"
    }
  }
}
matches {
  name: "in_port"
  optional {
    value {
      str: "7"
    }
  }
}
priority: 2030
action {
  name: "admit_to_l3"
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: matches[1].optional.value.str: "7" -> "a_port"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "acl_ingress_table"
matches {
  name: "is_ipv4"
  optional {
    value {
      hex_str: "0x1"
    }
  }
}
matches {
  name: "ip_protocol"
  ternary {
    value {
      hex_str: "0x11"
    }
    mask {
      hex_str: "0xff"
    }
  }
}
matches {
  name: "l4_dst_port"
  ternary {
    value {
      hex_str: "0x0223"
    }
    mask {
      hex_str: "0xffff"
    }
  }
}
matches {
  name: "in_port"
  optional {
    value {
      str: "8"
    }
  }
}
priority: 3100
action {
  name: "acl_drop"
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: matches[3].optional.value.str: "8" -> "2"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "acl_pre_ingress_table"
matches {
  name: "in_port"
  optional {
    value {
      str: "9"
    }
  }
}
priority: 1151
action {
  name: "set_vrf"
  params {
    name: "vrf_id"
    value {
      str: "vrf-210"
    }
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: matches[0].optional.value.str: "9" -> "1"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "acl_egress_table"
matches {
  name: "ether_type"
  ternary {
    value {
      hex_str: "0x0800"
    }
    mask {
      hex_str: ""
    }
  }
}
matches {
  name: "ip_protocol"
  ternary {
    value {
      hex_str: "0x11"
    }
    mask {
      hex_str: "0xff"
    }
  }
}
matches {
  name: "l4_dst_port"
  ternary {
    value {
      hex_str: "0x0223"
    }
    mask {
      hex_str: "0xffff"
    }
  }
}
matches {
  name: "out_port"
  optional {
    value {
      str: "10"
    }
  }
}
priority: 3101
action {
  name: "acl_drop"
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: matches[3].optional.value.str: "10" -> "2"

-- ORIGINAL ENTRY --------------------------------------------------------------
table_name: "wcmp_group_table"
matches {
  name: "wcmp_group_id"
  exact {
    str: "group-4294934545"
  }
}
action_set {
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-1"
        }
      }
    }
    weight: 1
    watch_port: "11"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "12"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "13"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "14"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "15"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "16"
  }
  actions {
    action {
      name: "set_nexthop_id"
      params {
        name: "nexthop_id"
        value {
          str: "nexthop-2"
        }
      }
    }
    weight: 1
    watch_port: "17"
  }
}
-- MODIFICATIONS ---------------------------------------------------------------
modified: action_set.actions[0].watch_port: "11" -> "1"
modified: action_set.actions[1].watch_port: "12" -> "a_port"
modified: action_set.actions[2].watch_port: "13" -> "2"
modified: action_set.actions[3].watch_port: "14" -> "1"
modified: action_set.actions[4].watch_port: "15" -> "a_port"
modified: action_set.actions[5].watch_port: "16" -> "2"
modified: action_set.actions[6].watch_port: "17" -> "1"
